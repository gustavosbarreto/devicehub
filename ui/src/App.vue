<template>
<v-app>
    <v-navigation-drawer v-if="isLoggedIn" :clipped="false" fixed v-model="drawer" :mini-variant="false" enable-resize-watcher app>
        <v-container>
            <div class="text-center">
                <v-icon>mdi-console</v-icon>
                <h2 style="font-family: monospace">ShellHub</h2>
                <span class="overline">beta</span>
            </div>
        </v-container>
        <v-list>
            <v-list-item v-for="item in items" :key="item.title" :to="item.path" two-line>
                <v-list-item-action>
                    <v-icon v-text="item.icon"></v-icon>
                </v-list-item-action>

                <v-list-item-content>
                    <v-list-item-title v-text="item.title"></v-list-item-title>
                </v-list-item-content>
            </v-list-item>
        </v-list>
    </v-navigation-drawer>
    <v-app-bar app elevate-on-scroll color="primary" class="pl-3 pr-4" v-if="isLoggedIn">
        <v-menu transition="scale-transition" origin="top left">
            <template v-slot:activator="{ on }">
                <v-chip v-on="on">
                    <v-icon left>mdi-server</v-icon>
                    My Device Fleet
                    <v-icon right>mdi-chevron-down</v-icon>
                </v-chip>
            </template>
        </v-menu>
        <v-spacer></v-spacer>
        <v-chip>
            <v-icon>help</v-icon>
        </v-chip>
        <v-chip>
            <v-icon>notifications</v-icon>
        </v-chip>
        <v-menu transition="scale-transition" origin="top right">
            <template v-slot:activator="{ on }">
                <v-chip v-on="on">
                    <v-icon left>mdi-account</v-icon>
                    {{ $store.getters["auth/currentUser"] }}
                    <v-icon right>mdi-chevron-down</v-icon>
                </v-chip>
            </template>

            <v-card>
                <v-list-item three-line>
                    <v-list-item-content>
                        <v-list-item-title class="mb-1">Tenant ID</v-list-item-title>
                        <v-list-item-subtitle>
                            <v-chip>
                                <span>
                                    {{ tenant }}
                                </span>
                                <v-icon right v-clipboard="tenant" v-clipboard:success="() => { copySnack = true; }">mdi-content-copy</v-icon>
                            </v-chip>
                        </v-list-item-subtitle>
                    </v-list-item-content>
                </v-list-item>

                <v-card-actions>
                    <v-btn small text @click="logout()">
                        Logout
                    </v-btn>
                </v-card-actions>
            </v-card>
        </v-menu>
    </v-app-bar>
    <v-content>
        <v-container class="pa-8" fluid>
            <router-view></router-view>
        </v-container>
        <v-snackbar v-model="copySnack" :timeout=3000>Tenant ID copied to clipboard</v-snackbar>
    </v-content>
</v-app>
</template>

<script>
export default {
  name: "App",

  methods: {
    logout() {
      this.$store.dispatch("auth/logout").then(() => {
        this.$router.push("/login");
      });
    }
  },

  computed: {
    tenant() {
      return this.$store.getters["auth/tenant"];
    },

    isLoggedIn() {
      return this.$store.getters["auth/isLoggedIn"];
    }
  },

  data() {
    return {
      drawer: true,
      clipped: false,
      copySnack: false,
      items: [
        {
          icon: "dashboard",
          title: "Dashboard",
          path: "/"
        },
        {
          icon: "devices",
          title: "Devices",
          path: "/devices"
        },
        {
          icon: "history",
          title: "Sessions",
          path: "/sessions"
        }
      ]
    };
  }
};
</script>

<style>
.v-list-active {
  border-left: 4px solid var(--v-primary-base);
}

.text-shadow {
  text-shadow: #000 0 0 6px;
  color: transparent;
}
</style>
