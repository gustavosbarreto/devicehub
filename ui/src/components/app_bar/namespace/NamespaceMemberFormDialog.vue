<template>
  <fragment>
    <v-tooltip
      v-if="addUser"
      bottom
      :disabled="hasAuthorization"
    >
      <template #activator="{ on }">
        <div v-on="on">
          <v-btn
            :disabled="!hasAuthorization"
            class="mr-2"
            outlined
            data-test="add-btn"
            @click="dialog = !dialog"
          >
            Add Member
          </v-btn>
        </div>
      </template>

      <span>
        You don't have this kind of authorization.
      </span>
    </v-tooltip>

    <v-tooltip
      v-else
      bottom
      :disabled="hasAuthorization"
    >
      <template #activator="{ on }">
        <span v-on="on">
          <v-btn
            :disabled="!hasAuthorization"
            class="mr-2"
            outlined
            data-test="edit-btn"
            @click="dialog = !dialog"
          >
            <v-icon
              outlined
              :disabled="!hasAuthorization"
              v-on="on"
            >
              edit
            </v-icon>
          </v-btn>
        </span>
      </template>

      <div>
        <span
          v-if="hasAuthorization"
          data-test="text-tooltip"
        >
          Edit
        </span>

        <span
          v-else
        >
          You don't have this kind of authorization.
        </span>
      </div>
    </v-tooltip>

    <v-dialog
      v-model="dialog"
      max-width="450"
      @click:outside="close"
    >
      <v-card data-test="namespaceNewMember-dialog">
        <v-card-title class="headline grey lighten-2 text-center">
          {{ addUser ? 'Add member to namespace' : 'Update member role' }}
        </v-card-title>

        <ValidationObserver
          ref="obs"
          v-slot="{ passes }"
        >
          <v-card-text class="caption mb-0">
            <ValidationProvider
              v-slot="{ errors }"
              ref="providerUsername"
              name="Username"
              rules="required"
              vid="username"
            >
              <v-text-field
                v-model="memberLocal.username"
                :disabled="!addUser"
                label="Username"
                :error-messages="errors"
                require
                data-test="username-text"
              />
            </ValidationProvider>
          </v-card-text>

          <v-card-text class="caption mb-0 pt-0">
            <ValidationProvider
              v-slot="{ errors }"
              ref="providerRole"
              vid="role"
              name="role"
              rules="required"
            >
              <v-row align="center">
                <v-col cols="12">
                  <v-select
                    v-model="memberLocal.selectedRole"
                    :items="items"
                    label="Role"
                    :error-messages="errors"
                    require
                    data-test="role-select"
                  />
                </v-col>
              </v-row>
            </ValidationProvider>
          </v-card-text>

          <v-card-actions>
            <v-spacer />
            <v-btn
              text
              data-test="dialogClose-btn"
              @click="close"
            >
              Close
            </v-btn>

            <v-btn
              v-if="addUser"
              color="primary"
              text
              data-test="dialogAdd-btn"
              @click="passes(addMember)"
            >
              Add
            </v-btn>

            <v-btn
              v-else
              color="primary"
              text
              data-test="dialogEdit-btn"
              @click="passes(editMember)"
            >
              Edit
            </v-btn>
          </v-card-actions>
        </ValidationObserver>
      </v-card>
    </v-dialog>
  </fragment>
</template>

<script>

import {
  ValidationObserver,
  ValidationProvider,
} from 'vee-validate';

import hasPermission from '@/components/filter/permission';

export default {
  name: 'NamespaceNewMemberComponent',

  filters: { hasPermission },

  components: {
    ValidationProvider,
    ValidationObserver,
  },

  props: {
    member: {
      type: Object,
      required: false,
      default: Object,
    },

    addUser: {
      type: Boolean,
      required: true,
    },
  },

  data() {
    return {
      dialog: false,
      username: '',
      selectedRole: '',
      memberLocal: [],
      items: ['administrator', 'operator', 'observer'],
    };
  },

  computed: {
    hasAuthorization() {
      const ownerID = this.$store.getters['namespaces/get'].owner;
      if (this.member.id === ownerID) {
        return false;
      }

      const role = this.$store.getters['auth/role'];
      if (role !== '') {
        let action = '';
        if (this.addUser) action = 'addMember';
        else action = 'removeMember';

        return hasPermission(
          this.$authorizer.role[role],
          this.$actions.namespace[action],
        );
      }

      return false;
    },
  },

  async created() {
    await this.setLocalVariable();
  },

  async updated() {
    await this.setLocalVariable();
  },

  methods: {
    setLocalVariable() {
      if (this.addUser) {
        this.memberLocal = {
          id: '',
          username: '',
          selectedRole: '',
        };
      } else {
        this.memberLocal = { ...this.member, selectedRole: this.member.role };
      }
    },

    async addMember() {
      try {
        await this.$store.dispatch('namespaces/addUser', {
          username: this.memberLocal.username,
          tenant_id: this.$store.getters['auth/tenant'],
          role: this.memberLocal.selectedRole,
        });

        this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.namespaceNewMember);
        this.update();
      } catch (error) {
        if (error.response.status === 404) {
          this.$refs.obs.setErrors({
            username: 'The username doesn\'t exist.',
          });
        } else if (error.response.status === 409) {
          this.$refs.obs.setErrors({
            username: 'The username has already been added to namespace.',
          });
        } else {
          this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.namespaceNewMember);
        }
      }
    },

    async editMember() {
      try {
        await this.$store.dispatch('namespaces/editUser', {
          user_id: this.memberLocal.id,
          tenant_id: this.$store.getters['auth/tenant'],
          role: this.memberLocal.selectedRole,
        });

        this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.namespaceEditMember);
        this.update();
      } catch (error) {
        if (error.response.status === 400) {
          this.$refs.obs.setErrors({
            username: 'The user isn\'t linked to the namespace.',
          });
        } else if (error.response.status === 403) {
          this.$refs.obs.setErrors({
            role: 'You don\'t have permission to assign a role to the user.',
          });
        } else if (error.response.status === 404) {
          this.$refs.obs.setErrors({
            username: 'The username doesn\'t exist.',
          });
        } else {
          this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.namespaceEditMember);
        }
      }
    },

    update() {
      this.$emit('update');
      this.close();
    },

    close() {
      this.dialog = false;
      this.$refs.obs.reset();
    },
  },
};

</script>
