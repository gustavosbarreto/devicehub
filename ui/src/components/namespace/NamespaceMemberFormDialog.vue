<template>
  <fragment>
    <v-tooltip
      v-if="addUser"
      bottom
      :disabled="hasAuthorization"
    >
      <template #activator="{ on }">
        <div v-on="on">
          <v-btn
            :disabled="!hasAuthorization"
            class="mr-2"
            color="primary"
            data-test="add-btn"
            @click="setShowDialog()"
          >
            Add Member
          </v-btn>
        </div>
      </template>

      <span>
        You don't have this kind of authorization.
      </span>
    </v-tooltip>

    <v-tooltip
      v-else
      bottom
      :disabled="hasAuthorization"
    >
      <template #activator="{ on }">
        <span v-on="on">
          <v-list-item-title data-test="edit-list">
            Edit
          </v-list-item-title>
        </span>

        <span v-on="on">
          <v-icon
            :disabled="!hasAuthorization"
            left
            data-test="edit-icon"
            v-on="on"
          >
            mdi-pencil
          </v-icon>
        </span>
      </template>

      <span>
        You don't have this kind of authorization.
      </span>
    </v-tooltip>

    <v-dialog
      v-model="showDialog"
      max-width="450"
      @click:outside="close"
    >
      <v-card data-test="namespaceNewMember-dialog">
        <v-card-title class="headline primary">
          {{ addUser ? 'Add member to namespace' : 'Update member role' }}
        </v-card-title>

        <ValidationObserver
          ref="obs"
          v-slot="{ passes }"
        >
          <v-card-text class="caption mb-0">
            <ValidationProvider
              v-slot="{ errors }"
              ref="providerUsername"
              name="Username"
              rules="required"
              vid="username"
            >
              <v-text-field
                v-model="memberLocal.username"
                :disabled="!addUser"
                label="Username"
                :error-messages="errors"
                require
                data-test="username-text"
              />
            </ValidationProvider>
          </v-card-text>

          <v-card-text class="caption mb-0 pt-0">
            <ValidationProvider
              v-slot="{ errors }"
              ref="providerRole"
              vid="role"
              name="role"
              rules="required"
            >
              <v-row align="center">
                <v-col cols="12">
                  <v-select
                    v-model="memberLocal.selectedRole"
                    :items="items"
                    label="Role"
                    :error-messages="errors"
                    require
                    data-test="role-select"
                  />
                </v-col>
              </v-row>
            </ValidationProvider>
          </v-card-text>

          <v-card-actions>
            <v-spacer />
            <v-btn
              text
              data-test="dialogClose-btn"
              @click="close()"
            >
              Close
            </v-btn>

            <v-btn
              v-if="addUser"
              color="primary"
              text
              data-test="dialogAdd-btn"
              @click="passes(addMember)"
            >
              Add
            </v-btn>

            <v-btn
              v-else
              color="primary"
              text
              data-test="dialogEdit-btn"
              @click="passes(editMember)"
            >
              Edit
            </v-btn>
          </v-card-actions>
        </ValidationObserver>
      </v-card>
    </v-dialog>
  </fragment>
</template>

<script>

import {
  ValidationObserver,
  ValidationProvider,
} from 'vee-validate';

import hasPermission from '@/components/filter/permission';

export default {
  name: 'NamespaceNewMemberComponent',

  filters: { hasPermission },

  components: {
    ValidationProvider,
    ValidationObserver,
  },

  props: {
    member: {
      type: Object,
      required: false,
      default: Object,
    },

    addUser: {
      type: Boolean,
      required: true,
    },

    show: {
      type: Boolean,
      required: false,
      default: false,
    },
  },

  data() {
    return {
      username: '',
      selectedRole: '',
      memberLocal: [],
      items: ['administrator', 'operator', 'observer'],
    };
  },

  computed: {
    showDialog: {
      get() {
        return this.show && this.hasAuthorization;
      },

      set(value) {
        this.$emit('show', value);
      },
    },

    hasAuthorization() {
      const ownerID = this.$store.getters['namespaces/get'].owner;
      if (this.member.id === ownerID) {
        return false;
      }

      const role = this.$store.getters['auth/role'];
      if (role !== '') {
        let action = '';
        if (this.addUser) action = 'addMember';
        else action = 'removeMember';

        return hasPermission(
          this.$authorizer.role[role],
          this.$actions.namespace[action],
        ) && this.member.role !== role;
      }

      return false;
    },
  },

  async created() {
    await this.setLocalVariable();
  },

  async updated() {
    await this.setLocalVariable();
  },

  methods: {
    setLocalVariable() {
      if (this.addUser) {
        this.memberLocal = {
          id: '',
          username: '',
          selectedRole: '',
        };
      } else {
        this.memberLocal = { ...this.member, selectedRole: this.member.role };
      }
    },

    async addMember() {
      try {
        await this.$store.dispatch('namespaces/addUser', {
          username: this.memberLocal.username,
          tenant_id: this.$store.getters['auth/tenant'],
          role: this.memberLocal.selectedRole,
        });

        this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.namespaceNewMember);
        this.update();
      } catch (error) {
        if (error.response.status === 404) {
          this.$refs.obs.setErrors({
            username: 'The username doesn\'t exist.',
          });
        } else if (error.response.status === 409) {
          this.$refs.obs.setErrors({
            username: 'The username has already been added to namespace.',
          });
        } else {
          this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.namespaceNewMember);
        }
      }
    },

    async editMember() {
      try {
        await this.$store.dispatch('namespaces/editUser', {
          user_id: this.memberLocal.id,
          tenant_id: this.$store.getters['auth/tenant'],
          role: this.memberLocal.selectedRole,
        });

        this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.namespaceEditMember);
        this.update();
      } catch (error) {
        if (error.response.status === 400) {
          this.$refs.obs.setErrors({
            username: 'The user isn\'t linked to the namespace.',
          });
        } else if (error.response.status === 403) {
          this.$refs.obs.setErrors({
            role: 'You don\'t have permission to assign a role to the user.',
          });
        } else if (error.response.status === 404) {
          this.$refs.obs.setErrors({
            username: 'The username doesn\'t exist.',
          });
        } else {
          this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.namespaceEditMember);
        }
      }
    },

    update() {
      this.$emit('update');
      this.close();
    },

    setShowDialog() {
      this.$emit('update:show', true);
    },

    close() {
      this.$emit('update:show', false);
      this.$refs.obs.reset();
    },
  },
};

</script>
