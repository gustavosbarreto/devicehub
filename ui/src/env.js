export default {
  /* eslint-disable no-param-reassign */
  install(Vue) {
    const env = {
      isHosted: window.env.HOSTED === 'true',
    };

    Vue.env = env;
    Vue.prototype.$env = env;
  },
  /* eslint-enable no-param-reassign */
};
