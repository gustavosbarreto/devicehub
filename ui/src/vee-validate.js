import {
  required,
  integer,
  email,
  confirmed,
} from 'vee-validate/dist/rules';
import { extend } from 'vee-validate';
import isValidHostname from 'is-valid-hostname';

import { parsePrivateKey, parseKey } from '@/sshpk';

extend('required', {
  ...required,
  message: 'This field is required',
});

extend('integer', {
  ...integer,
  message: 'This value must be a integer number',
});

extend('email', {
  ...email,
  message: 'This field must be a valid email',
});

extend('noDot', {
  validate: (value) => !/\./.test(value),
  message: 'The name must not contain dots',
});

extend('rfc1123', {
  validate: (value) => isValidHostname(value),
  message: 'You entered an invalid RFC1123 name',
});

extend('routeIdentifier', {
  validate: (value) => !/\/|@|&|:/.test(value),
  message: 'The name must not contain /, @, &, and :.',
});

extend('password', (value) => {
  if (value.length < 5 || value.length > 30) {
    return 'Your password should be 5-30 characters long';
  }
  return true;
});

extend('namespace', (value) => {
  if (value.length < 3 || value.length > 30) {
    return 'Your namespace should be 3-30 characters long';
  }
  return true;
});

extend('device', (value) => {
  if (value.length < 3 || value.length > 30) {
    return 'Your hostname should be 3-30 characters long';
  }
  return true;
});

extend('tag', (value) => {
  if (value.length < 3 || value.length > 255) {
    return 'Your tag should be 3-255 characters long';
  }
  return true;
});

extend('tagRequired', {
  ...required,
  message: 'You must choose at least one tag',
});

extend('tagsLength', {
  validate(value) {
    if (value.length > 3) {
      return 'The maximum capacity has reached';
    }
    return true;
  },
});

extend('comparePasswords', {
  validate(value, { currentPassword }) {
    if (value === currentPassword) {
      return false;
    }
    return true;
  },
  params: ['currentPassword'],
  message: 'The passwords are the same',
});

extend('confirmed', {
  ...confirmed,
  message: 'The passwords do not match',
});

extend('parseKey', {
  validate(value, args) {
    try {
      if (args.typeKey === 'private') {
        parsePrivateKey(value);
      } else {
        parseKey(value);
      }

      return true;
    } catch (err) {
      return false;
    }
  },
  params: ['typeKey'],
  message: 'Not valid key',
});
