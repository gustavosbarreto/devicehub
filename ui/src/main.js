import Vue from 'vue';

import Snackbar from '@/components/snackbar/Snackbar';

import App from './App';
import router from './router';
import store from './store';
import env from './env';
import './vee-validate';
import vuetify from './plugins/vuetify';

Vue.config.productionTip = false;

Vue.component('Snackbar', Snackbar);

Vue.use(require('vue-moment'));

Vue.use(env);

new Vue({
  vuetify,
  router,
  store,
  render: (h) => h(App),
}).$mount('#app');
