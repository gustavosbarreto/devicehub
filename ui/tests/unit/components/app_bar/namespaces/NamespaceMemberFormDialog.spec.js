import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import { ValidationProvider, ValidationObserver } from 'vee-validate';
import flushPromises from 'flush-promises';
import Vuetify from 'vuetify';
import NamespaceMemberFormDialog from '@/components/app_bar/namespace/NamespaceMemberFormDialog';
import { actions, authorizer } from '../../../../../src/authorizer';
import '@/vee-validate';

describe('NamespaceMemberFormDialog', () => {
  const localVue = createLocalVue();
  const vuetify = new Vuetify();
  localVue.use(Vuex);
  localVue.component('ValidationProvider', ValidationProvider);
  localVue.component('ValidationObserver', ValidationObserver);

  document.body.setAttribute('data-app', true);

  let wrapper;

  const accessType = ['owner', 'administrator', 'operator', 'observer'];

  const hasAuthorization = {
    owner: true,
    administrator: true,
    operator: false,
    observer: false,
  };

  const members = [
    {
      id: 'xxxxxxxx',
      type: 'owner',
      username: 'user1',
    },
    {
      id: 'xxxxxxxy',
      type: 'observer',
      username: 'user2',
    },
  ];

  const memberLocal = {
    id: '',
    selectedAccessType: '',
    username: '',
  };

  const memberLocalEdit = {
    id: 'xxxxxxxy',
    type: 'observer',
    selectedAccessType: 'observer',
    username: 'user2',
  };

  const namespaceGlobal = {
    name: 'namespace',
    owner: 'user1',
    members,
    tenant_id: 'xxxxxxxx',
    devices_count: 0,
    max_devices: 0,
  };

  const tests = [
    {
      description: 'Icon add user',
      variables: {
        dialog: false,
        namespaceGlobal,
      },
      props: {
        member: {},
        addUser: true,
      },
      data: {
        dialog: false,
        username: '',
        selectedAccessType: '',
        memberLocal,
        items: ['administrator', 'operator', 'observer'],
      },
      template: {
        'add-btn': true,
        'edit-btn': false,
        'namespaceNewMember-dialog': false,
        'dialogClose-btn': false,
        'dialogAdd-btn': false,
        'dialogEdit-btn': false,
      },
    },
    {
      description: 'Icon edit user',
      variables: {
        namespaceGlobal,
        dialog: false,
      },
      props: {
        member: members[1],
        addUser: false,
      },
      data: {
        dialog: false,
        username: '',
        selectedAccessType: '',
        memberLocal: memberLocalEdit,
        items: ['administrator', 'operator', 'observer'],
      },
      template: {
        'add-btn': false,
        'edit-btn': true,
        'namespaceNewMember-dialog': false,
        'dialogClose-btn': false,
        'dialogAdd-btn': false,
        'dialogEdit-btn': false,
      },
    },
    {
      description: 'Dialog create user',
      variables: {
        dialog: true,
        namespaceGlobal,
      },
      props: {
        member: {},
        addUser: true,
      },
      data: {
        dialog: true,
        username: '',
        selectedAccessType: '',
        memberLocal,
        items: ['administrator', 'operator', 'observer'],
      },
      template: {
        'add-btn': true,
        'edit-btn': false,
        'namespaceNewMember-dialog': true,
        'dialogClose-btn': true,
        'dialogAdd-btn': true,
        'dialogEdit-btn': false,
      },
    },
    {
      description: 'Dialog edit user',
      variables: {
        namespaceGlobal,
        dialog: true,
      },
      props: {
        member: members[1],
        addUser: false,
      },
      data: {
        dialog: true,
        username: '',
        selectedAccessType: '',
        memberLocal: memberLocalEdit,
        items: ['administrator', 'operator', 'observer'],
      },
      template: {
        'add-btn': false,
        'edit-btn': true,
        'namespaceNewMember-dialog': true,
        'dialogClose-btn': true,
        'dialogAdd-btn': false,
        'dialogEdit-btn': true,
      },
    },
  ];

  const storeVuex = (namespace, currentAccessType) => new Vuex.Store({
    namespaced: true,
    state: {
      namespace,
      currentAccessType,
    },
    getters: {
      'namespaces/get': (state) => state.namespace,
      'auth/accessType': (state) => state.currentAccessType,
    },
    actions: {
      'namespaces/adduser': () => {},
      'snackbar/showSnackbarSuccessAction': () => {},
      'snackbar/showSnackbarErrorAction': () => {},
    },
  });

  tests.forEach((test) => {
    accessType.forEach((currentAccessType) => {
      describe(`${test.description} ${currentAccessType}`, () => {
        beforeEach(() => {
          wrapper = mount(NamespaceMemberFormDialog, {
            store: storeVuex(test.variables.namespaceGlobal, currentAccessType),
            localVue,
            stubs: ['fragment'],
            propsData: { member: test.props.member, addUser: test.props.addUser },
            vuetify,
            mocks: {
              $authorizer: authorizer,
              $actions: actions,
            },
          });

          wrapper.setData({ dialog: test.variables.dialog });
        });

        ///////
        // Component Rendering
        //////

        it('Is a Vue instance', () => {
          expect(wrapper).toBeTruthy();
        });
        it('Renders the component', () => {
          expect(wrapper.html()).toMatchSnapshot();
        });

        ///////
        // Data checking
        //////

        it('Receive data in props', () => {
          Object.keys(test.props).forEach((item) => {
            expect(wrapper.vm[item]).toEqual(test.props[item]);
          });
        });
        it('Compare data with default value', () => {
          Object.keys(test.data).forEach((item) => {
            expect(wrapper.vm[item]).toEqual(test.data[item]);
          });
        });
        it('Process data in the computed', () => {
          expect(wrapper.vm.hasAuthorization).toEqual(hasAuthorization[currentAccessType]);
        });

        //////
        // HTML validation
        //////

        it('Renders the template with data', () => {
          Object.keys(test.template).forEach((item) => {
            expect(wrapper.find(`[data-test="${item}"]`).exists()).toBe(test.template[item]);
          });
        });

        if (test.data.dialog) {
          if (hasAuthorization[currentAccessType] && !test.props.addUser) {
            it('Show validation messages', async () => {
              wrapper.setData({ memberLocal: { username: '' } });
              await flushPromises();

              const validator = wrapper.vm.$refs.providerUsername;

              await validator.validate();
              expect(validator.errors[0]).toBe('This field is required');
            });
          }
        }
      });
    });
  });
});
