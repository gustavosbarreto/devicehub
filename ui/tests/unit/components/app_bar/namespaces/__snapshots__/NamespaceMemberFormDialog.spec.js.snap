// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NamespaceMemberFormDialog Dialog create user administrator Renders the component 1`] = `
<fragment-stub>
  <div><button type="button" class="mr-2 v-btn v-btn--outlined theme--light v-size--default" data-test="add-btn"><span class="v-btn__content">
          Add Member
        </span></button></div><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Dialog create user observer Renders the component 1`] = `
<fragment-stub>
  <div><button type="button" disabled="disabled" class="mr-2 v-btn v-btn--disabled v-btn--outlined theme--light v-size--default" data-test="add-btn"><span class="v-btn__content">
          Add Member
        </span></button></div><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Dialog create user operator Renders the component 1`] = `
<fragment-stub>
  <div><button type="button" disabled="disabled" class="mr-2 v-btn v-btn--disabled v-btn--outlined theme--light v-size--default" data-test="add-btn"><span class="v-btn__content">
          Add Member
        </span></button></div><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Dialog create user owner Renders the component 1`] = `
<fragment-stub>
  <div><button type="button" class="mr-2 v-btn v-btn--outlined theme--light v-size--default" data-test="add-btn"><span class="v-btn__content">
          Add Member
        </span></button></div><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Dialog edit user administrator Renders the component 1`] = `
<fragment-stub><span><button type="button" class="mr-2 v-btn v-btn--outlined theme--light v-size--default" data-test="edit-btn"><span class="v-btn__content"><i aria-hidden="true" outlined="" class="v-icon notranslate material-icons theme--light">edit</i></span></button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Dialog edit user observer Renders the component 1`] = `
<fragment-stub><span><button type="button" disabled="disabled" class="mr-2 v-btn v-btn--disabled v-btn--outlined theme--light v-size--default" data-test="edit-btn"><span class="v-btn__content"><i aria-hidden="true" outlined="" class="v-icon notranslate v-icon--disabled material-icons theme--light">edit</i></span></button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Dialog edit user operator Renders the component 1`] = `
<fragment-stub><span><button type="button" disabled="disabled" class="mr-2 v-btn v-btn--disabled v-btn--outlined theme--light v-size--default" data-test="edit-btn"><span class="v-btn__content"><i aria-hidden="true" outlined="" class="v-icon notranslate v-icon--disabled material-icons theme--light">edit</i></span></button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Dialog edit user owner Renders the component 1`] = `
<fragment-stub><span><button type="button" class="mr-2 v-btn v-btn--outlined theme--light v-size--default" data-test="edit-btn"><span class="v-btn__content"><i aria-hidden="true" outlined="" class="v-icon notranslate material-icons theme--light">edit</i></span></button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Icon add user administrator Renders the component 1`] = `
<fragment-stub>
  <div><button type="button" class="mr-2 v-btn v-btn--outlined theme--light v-size--default" data-test="add-btn"><span class="v-btn__content">
          Add Member
        </span></button></div><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Icon add user observer Renders the component 1`] = `
<fragment-stub>
  <div><button type="button" disabled="disabled" class="mr-2 v-btn v-btn--disabled v-btn--outlined theme--light v-size--default" data-test="add-btn"><span class="v-btn__content">
          Add Member
        </span></button></div><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Icon add user operator Renders the component 1`] = `
<fragment-stub>
  <div><button type="button" disabled="disabled" class="mr-2 v-btn v-btn--disabled v-btn--outlined theme--light v-size--default" data-test="add-btn"><span class="v-btn__content">
          Add Member
        </span></button></div><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Icon add user owner Renders the component 1`] = `
<fragment-stub>
  <div><button type="button" class="mr-2 v-btn v-btn--outlined theme--light v-size--default" data-test="add-btn"><span class="v-btn__content">
          Add Member
        </span></button></div><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Icon edit user administrator Renders the component 1`] = `
<fragment-stub><span><button type="button" class="mr-2 v-btn v-btn--outlined theme--light v-size--default" data-test="edit-btn"><span class="v-btn__content"><i aria-hidden="true" outlined="" class="v-icon notranslate material-icons theme--light">edit</i></span></button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Icon edit user observer Renders the component 1`] = `
<fragment-stub><span><button type="button" disabled="disabled" class="mr-2 v-btn v-btn--disabled v-btn--outlined theme--light v-size--default" data-test="edit-btn"><span class="v-btn__content"><i aria-hidden="true" outlined="" class="v-icon notranslate v-icon--disabled material-icons theme--light">edit</i></span></button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Icon edit user operator Renders the component 1`] = `
<fragment-stub><span><button type="button" disabled="disabled" class="mr-2 v-btn v-btn--disabled v-btn--outlined theme--light v-size--default" data-test="edit-btn"><span class="v-btn__content"><i aria-hidden="true" outlined="" class="v-icon notranslate v-icon--disabled material-icons theme--light">edit</i></span></button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`NamespaceMemberFormDialog Icon edit user owner Renders the component 1`] = `
<fragment-stub><span><button type="button" class="mr-2 v-btn v-btn--outlined theme--light v-size--default" data-test="edit-btn"><span class="v-btn__content"><i aria-hidden="true" outlined="" class="v-icon notranslate material-icons theme--light">edit</i></span></button></span><span class="v-tooltip v-tooltip--bottom"><!----></span>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;
